include/url.h

```diff
@@ -0,0 +1,86 @@
+#ifndef URL_CPP_H
```

Comment:
Is there a particular reason you are using a struct and not a class?  I believe what you really want here is a class.